### DISCUSSION TITLE
Guess the number application

### DISCUSSION THREAD

**Discussion by Yoy017** [heig-vd-dai-course/heig-vd-dai-course#468] [Q&A] [answered]
Bonjour,

J'ai recréer l'application guess the number du cours pratique mais je n'arrive pas à faire une chose. Je voudrais mettre dans mon Dockerfile `CMD ["java", "-jar", "/app/java-tcp-programming-1.0-SNAPSHOT.jar", "client", "-H", "localhost"]` ainsi le jeu ce lance dès qu'on run l'image docker.
![image](https://github.com/user-attachments/assets/603dcb4f-150f-4474-b8ff-de25ce7437d7)

Le soucis c'est que lorsqu'on lance le serveur, il se met en écoute et on ne peut pas entrer une autre commande. Comment je peux lancer le serveur en background (je ne peux pas ajouté -d dans la ligne de commande) pour pouvoir lancer le jeu ?

**Comment by Yoy017** [comment]
Enfaite, j'ai compris comment vous vouliez que j'utilise l'application. Je rectifie mon problème, lorsque je veux run l'image côté client lorsque le serveur (donc le port) est ouvert je me retrouve avec l'erreur suivante
![image](https://github.com/user-attachments/assets/b6bbf56f-4b21-42ce-9416-8333ecbc3421)

**Reply by ludelafo** [reply]
Bonjour @Yoy017,

Merci pour votre message et déjà bravo d'avoir réussi à implémenter le jeu _"Guess the number"_ !

Pour répondre à votre question, la première commande que vous avez exécutée a lancé un container du serveur en arrière-plan portant le nom `the-server`. Comme indiqué dans le support de cours [ici](https://github.com/heig-vd-dai-course/heig-vd-dai-course/blob/main/12-java-tcp-programming/COURSE_MATERIAL.md#dockerize-the-application), l'argument `--name` de Docker définit le nom du container ainsi que le hostname de celui-ci. Cela va permettre au client de pouvoir accéder au serveur en utilisant le nom DNS `the-server`.

Votre serveur semble avoir donc bien démarré et attend des requêtes de la part de votre client. Pour s'assurer que tout a vraiment bien été lancé, j'exécuterais le serveur sans le `-d` pour voir ses logs et lancer un second terminal pour lancer le client.

La dernière commande que vous avez lancée a pour but de lancer le client. Vous n'avez pas besoin de spécifier à nouveau l'argument `--name` de Docker, car celui-ci est utilisé uniquement pour le serveur. Vous pouvez donc lancer le serveur avec la commande `docker run --rm -it <docker-image-tag> client --host the-server` (tiré du support de cours). Notez que l'hôte n'est pas `localhost` mais `the-server`.

Avec Docker, `localhost` fait référence au container actuel. Donc si vous lancez votre client dans un environnement Docker avec comme hôte `localhost`, vous faites référence "à vous-même" (au container), ce qui est toujours un peu confus et à prendre en main dans les premières expériences avec Docker (c'est un peu "à la Matrix" mais on finit par bien comprendre).

Pensez donc à bien lancer votre client avec comme serveur hôte `the-server` plutôt que `localhost` et, grâce à l'argument Docker `--name` expliqué précédemment que vous avez spécifié pour le serveur, votre client devrait pouvoir contacter le serveur sans souci.

Faites-moi savoir si vous avez pu résoudre votre problème mais vous me semblez déjà bien parti !

Une toute bonne journée,
Ludovic

**Reply by Yoy017** [reply]
Merci je tente de trouver la solution. Je vous tiens au courant !

**Reply by Yoy017** [reply]
Bonjour, des nouvelles concernant le network ?

**Reply by ludelafo** [reply, marked as answer]
Bonjour Yoann,

Excusez-moi pour le temps de réponse, je venais de mettre à jour l'issue que j'utilise pour les points à améliorer pour l'année prochaine concernant ce point juste avant que vous ne m'écriviez : <https://github.com/heig-vd-dai-course/heig-vd-dai-course/issues/306> (je vais aussi mentionner cette même discussion comme inspiration pour la mise à jour du cours).

Je vous ai dit des bêtises dans mon dernier message en confondant et mélangeant des fonctionnalités de Docker et d'autres fonctionnalités de Docker Compose, qui lui gère automatiquement en partie les réseaux entre containers.

Pour corriger mes propos : chaque container démarré avec Docker (sans la partie Docker Compose) est indépendant les uns des autres. Il s'agit d'une mesure de sécurité et d'intégrité des containers : sans le faire explicitement, un container ne peut pas avoir accès à un autre (que ce soit niveau réseau ou stockage). Il est donc nécessaire de créer un réseau pour relier les containers pour qu'ils puissent communiquer comme vous l'avez fait.

Pour rappel, il est donc nécessaire de créer un réseau Docker comme suit :

```sh
# Create a Docker network
docker network create heig-vd-dai
```

Si besoin, il est possible de lister tous les réseaux Docker avec la commande suivante :

```sh
# List of Docker networks
docker network list
```

Et en supprimer avec la commande suivante :

```sh
# Delete a Docker network
docker network rm <name of the Docker network>
```

Une fois un réseau Docker créé, il est possible de démarrer les deux containers en les connectant sur ce réseau créé spécialement pour eux :

```sh
# Start the server on the Docker network with a hostname `the-server`
docker run --rm -it --network heig-vd-dai --name the-server <docker-image-tag> server

# Start the client on the Docker network and access the server using its hostname `the-server`
docker run --rm -it --network heig-vd-dai <docker-image-tag> client --host the-server
```

J'ai pu tester localement et tout semble fonctionner correctement de mon côté avec ces instructions.

Je vais essayer de mettre à jour les supports de cours au plus vite pour refléter cette discussion.

Merci pour votre patience et redites-moi si ça a aussi fonctionné chez vous !

**Reply by Yoy017** [reply]
Oui, je lance donc en créant mon propre network. Merci d'avoir répondu !
